# Домашнее задание №3

## Настройка среды

In [2]:
from huggingface_hub import notebook_login

notebook_login() # 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

## Глобальные параметры работы блокнота

In [32]:
# Выбранная контрольная точка модели
model_checkpoint = "google-t5/t5-small"

## Задание

[Ссылка](https://tn-gvu.mckx.ru/c/c4YcAACAurcBAEAA/sE43BQ/ZYcLPOYT_IPW6LK3/?u=https%3A%2F%2Fcontest.yandex.ru%2Fcontest%2F67637%2Fenter%2F%3Futm_source%3Dmindbox%26utm_medium%3Demail%26utm_campaign%3Dtraining6%26utm_content%3Ddigest%26utm_term%3D06.11.2024) на контест с заданием.

Необходимо обучить модель перевода с языка зетан на английский.

***Данные:*** https://disk.yandex.ru/d/u8mmcUBn64p_Nw

***Формат решения:*** json-lines в формате, аналогичной валидации, с переводами исходных текстов в тестовой выборке

***Метрика качества:*** BLEU между переводами и английскими референсами на закрытом тесте

## Загрузка данных

In [ ]:
# ! wget https://disk.yandex.ru/d/u8mmcUBn64p_Nw

In [3]:
from datasets import load_dataset, Features, Value

data_files = {
    "train": "train.jsonl",
    "validation": "val.jsonl",
    "test": "test_no_reference.jsonl"
}

features = Features({
    'src': Value('string'),
    'dst': Value('string'),
})

dataset = load_dataset("json", data_files=data_files, features=features)

Посмотрим на один пример данных из каждой части датасета:

In [4]:
dataset['train'][4]

{'src': '◈◠ ◧▱◠▦ ◀◫◓ ▨◠◉ ◂▱◠▽◈◠▦ ◀◠▷◞◪◈◗◳◧◓■ ◉◧◐▾▦▱◨◐▾ ○▱◎◠▦▱◠◓◈◠▦ ▨◠◉◠▦ ▽◠▷◨◈◫▱▴◓▵',
 'dst': "He's talking about a few right here in Lisbon, mostly Jews who have escaped the Germans, that's all."}

In [5]:
dataset['validation'][4]

{'src': '◲◒▱▴▫◗▱◪▦ ▻▾◀ ◚◪ ◀◠◓▱◠◓◈◠ ◀◨ ◠◳ ◳◫◳▴▼◪▨ ◞◠▫◬◒▱◠◓▪ ▽◭▢◈◪ ◭◉ ◈▩◒▴◓▨◪▦ ◗◉▨◫ ◞◠▫◬◒▱◠◓▪ ◳▩▢◈▴ 6■6 ◠◓▫▫▪▵"',
 'dst': 'Sales of drinks in pubs and bars increased by 6.6% over the month, while sales of food increased by 3%."'}

In [6]:
dataset['test'][4]

{'src': '"○◐▱◠◈◬◐▪▦▪ ◕◣◓◎◪▱◪◓◗▦◪ ◠◞▱◠ ◗▢◫▦ ◚▴◓◎▴■" ◈◪◈◫ ◀◠▦◠▵', 'dst': None}

Немного узнать об этом чудном языке не помешает. Посмотрим сколько уникальных символов в этом языке:

In [7]:
# Функция для подсчета уникальных символов
def count_unique_characters(dataset, feature_name):
    unique_characters = set()
    for split in dataset.keys():
        for example in dataset[split][feature_name]:
            if example is not None:
                unique_characters.update(example)
    return len(unique_characters), unique_characters

# Подсчет уникальных символов во всех частях датасета
num_unique_chars, unique_chars = count_unique_characters(dataset, "src")

print(f"Количество уникальных символов: {num_unique_chars}")
print(f"Уникальные символы: {unique_chars}")

Количество уникальных символов: 182
Уникальные символы: {'◂', ';', '=', '◨', ']', '◜', '◖', '◭', '—', '◀', 'è', 'ä', 'Î', '\x99', '◕', '▾', '▻', '◳', '~', '^', 'Ä', '◞', '[', '◉', '◆', ')', '◃', '▷', '\x9e', '\u200b', '¿', '◙', '§', '◱', '_', '‘', '◄', '8', '+', '▶', 'ë', 'É', '/', '‚', '◎', '▰', '▿', 'â', 'î', 'Þ', '▩', '▢', '`', '◤', '▽', '▼', '◩', 'Â', '™', 'º', 'é', '►', '▦', 'ø', '◥', '◯', '◪', '○', '£', '△', '5', '\x9d', '■', '▯', '▥', 'í', '\\', '◚', '▮', '◬', '♪', '◰', '0', ':', '6', '◧', '◛', '□', '¶', '◠', '▨', '7', '"', '{', 'ú', '½', '◮', '●', '▧', 'á', '◒', 'đ', '◅', 'ñ', '´', '◫', '4', '◘', '¡', '▹', '◡', '◁', '*', '$', '%', 'å', '◲', '–', '◦', '3', '▸', '◑', '◟', '’', 'ă', '◗', '◓', '&', '♫', '▣', "'", '@', '°', 'ο', 'ß', 'ð', '◐', '◍', '“', '•', 'ï', 'ô', '”', '▲', 'þ', ' ', 'ţ', '◢', '¤', 'ι', '▫', '─', '▭', '◔', '!', '◊', '▴', '▵', '?', '}', '9', '▪', '▬', 'ª', '◇', '◌', '\u202d', '\xa0', 'û', 'ý', '▤', '1', '2', '◈', '◣', '#', 'Ý', '◝', 'ν', 'ó', '▱', '('}


Посмотрим аналогичную статистику для целевых последовательностей:

In [8]:
# Подсчет уникальных символов во всех частях датасета для целевых последовательностей
num_unique_chars_dst, unique_chars_dst = count_unique_characters(dataset, "dst")

print(f"Количество уникальных символов в целевых последовательностях: {num_unique_chars_dst}")
print(f"Уникальные символы в целевых последовательностях: {unique_chars_dst}")

Количество уникальных символов в целевых последовательностях: 201
Уникальные символы в целевых последовательностях: {';', '鈥', 'Æ', 'ã', '=', '′', 'с', ']', 'Ã', 'H', '—', 'è', 'c', 'C', 'ä', 'š', '\x99', '€', 'R', 'æ', '~', 'k', '^', 'Ä', 'ư', 'fi', '[', 'À', 'I', 'ρ', ')', 'Ż', 'Ë', 'P', 'ö', '¿', '§', 'G', '_', 'fl', 'Á', 'l', 't', 'n', '8', '+', 'É', 'Β', 'ë', 'e', '/', 'f', '‚', 'u', 'â', 'Ο', 'S', '`', 'ò', 'Â', 'N', 'Z', 'J', '™', 'm', 'é', 'r', 'ø', 'B', '³', '£', 'İ', ',', '5', '\x9d', 'Η', 'í', '\\', '♪', '\x8d', '│', '0', 'a', ':', '6', 'D', '±', '¶', '7', '{', '"', 'A', '.', 'ć', '-', 'ú', 'z', '½', 'Μ', 'á', '●', 'T', 'X', 'w', 'ñ', '´', '\x97', '4', 'à', 'y', 'υ', '¡', 'E', 'Ν', 'M', '$', '*', '%', 'ί', 'L', 'å', '–', '©', '3', '\xad', '‒', 'õ', 'ü', 'Y', 'O', 'ă', 'Τ', 'd', '\x90', 'ş', 'œ', '\x83', '&', 'U', 'Ü', '♫', "'", '@', 'x', 'v', 'ç', '°', 'ο', 'p', '¬', 'ð', 'F', 'K', '“', 'ï', 'ô', '”', 'þ', ' ', '¤', 'j', 'g', '¢', '¯', '─', 'ê', '檛', 'Ι', 'Ö', '!', 'Q', 'W', '

Посмотрим на максимальную длинну строк для каждой части датасета:

In [20]:
max_lengths = {}

for split in dataset:
    src_lengths = [len(item["src"]) for item in dataset[split] if item["src"] is not None]
    dst_lengths = [len(item["dst"]) for item in dataset[split] if item["dst"] is not None]
    
    max_src_length = max(src_lengths) if src_lengths else 0
    max_dst_length = max(dst_lengths) if dst_lengths else 0
    
    max_lengths[split] = {"src": max_src_length, "dst": max_dst_length}

print(max_lengths)

{'train': {'src': 321, 'dst': 395}, 'validation': {'src': 504, 'dst': 352}, 'test': {'src': 458, 'dst': 0}}


Самая длинная строка присутствующая в датасете имеет длинну 504 символа.

## Подготовка данных

### Токенизатор

#### Нормализация, размер словаря

Нормализация включает в себя некоторую общую очистку, например, удаление ненужных пробельных символов, понижение регистра и/или удаление ударений. Посмотрим как выполняется нормализация токенизатором выбранной нами модели:

Для начала выберем модель, которую будет использовать для дообучения. Попробуем использовать модель :

In [33]:
from transformers import AutoTokenizer
example_num = 13

tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)



# Проверка нормализации текста
text = dataset['validation'][example_num]['src']
translate = dataset['validation'][example_num]['dst']

print("Оригинальный текст:", text)
print("Нормализованный текст:", tokenizer.backend_tokenizer.normalizer.normalize_str(text))
print("Перевод:", translate)

# Проверка, является ли токенизатор быстрым
print("Токенизатор быстрый:", tokenizer.is_fast)

# Узнаем текущий размер словаря токенизатора
current_vocab_size = tokenizer.vocab_size
print(f"Текущий размер словаря токенизатора: {current_vocab_size}")


Оригинальный текст: ▲◳◈▦▴◳'◈◪ ▽◠◒◠◳◠▦ 45 ▽◠◒▪▦◈◠▨◗ ◝◠◳◠▦ ◙◠◚◫◞■ ◧◐▱▾▦◨▦ ◕▴▱▴▼▴▨ ◚◠◠▫ ◪◈◪▦ ◚◪ ▴▫◓◠◍◬▦◈◠▨◫ ▷◪◓▨▴◞◗ ◕◭▱▩◎◞▴▫◪▦ ◀◗◓ ◠◒◉▪ ◂▱◈▾◐▾▦▾ ◞◇◳▱▩▽◂◓▵
Нормализованный текст: ▲◳◈▦▴◳'◈◪ ▽◠◒◠◳◠▦ 45 ▽◠◒▪▦◈◠▨◗ ◝◠◳◠▦ ◙◠◚◫◞■ ◧◐▱▾▦◨▦ ◕▴▱▴▼▴▨ ◚◠◠▫ ◪◈◪▦ ◚◪ ▴▫◓◠◍◬▦◈◠▨◫ ▷◪◓▨▴◞◗ ◕◭▱▩◎◞▴▫◪▦ ◀◗◓ ◠◒◉▪ ◂▱◈▾◐▾▦▾ ◞◇◳▱▩▽◂◓▵
Перевод: Davis, 45, who lives in Leadney, said her son was a great cook and had a charming smile.
Токенизатор быстрый: True
Текущий размер словаря токенизатора: 32100


Судя по выводу текст подвергается минимальной нормализации, значит проблем с потерей символов быть не должно. Размер предобученного словаря токенизатора для выбранной модели равен 32100 токен. 

#### Размер контекстного окна

In [ ]:
# Получим максимальную длину последовательности (размер контекстного окна) модели 
max_length = tokenizer.model_max_length

print(f"Максимальная длина последовательности модели: {max_length}")

Максимальная длина последовательности модели: 512


Максимальная длинна контекста для выбранной модели равна 512 токенов.

> ***Длина контекста (или контекстное окно)*** это максимальное количеству токенов, которые модель может обрабатывать одновременно.

#### Обучение нового токенизатора

Так как язык зетан является вымешленным (новым), необходимо обучить новый токенизатор:

In [24]:
def batch_iterator(batch_size=10):
    for split in dataset:
        batch_size = batch_size // 2
        for i in range(0, len(dataset[split]), batch_size):
            src_list = dataset[split][i : i + batch_size]["src"]
            dst_list = dataset[split][i : i + batch_size]["dst"]
            yield [str(src) for src in src_list] + [str(dst) for dst in dst_list]


vocab_size = current_vocab_size  # Можно начать с 32000 и экспериментировать

# Обучение токенизатора
new_tokenizer = tokenizer.train_new_from_iterator(batch_iterator(batch_size=1000), vocab_size=vocab_size)
print("Токенизатор обучен")




Токенизатор обучен


Проверим работу нового токенизатора на нескольких примерах исходного и целевого текста:

In [25]:
example_num = 1

# Получаем строки из датасета
text = dataset['validation'][example_num]['src']
translate = dataset['validation'][example_num]['dst']

# Токенизируем текст и перевод, выводим результат
print(text)
print(new_tokenizer(text), end="\n\n")

print(translate)
print(new_tokenizer(translate), end="\n\n")

print(new_tokenizer.vocab)

◤◪▦◫ ▨◠▦◞▴◓ ◠◒▪◞▪■ ◀◠◐▪◒◬▨▱▪▨ ◞◫◞▫◪◎◫▦▴ ▨◣▫◭ ▦◫◳◪▫▱◗ ▷▩▼◓◪▱▴◓◫ "◕◣◓◎▴▽◫" ◇◐◓◪▫▴◀◫▱◗◓
{'input_ids': [10743, 18257, 16672, 11029, 14803, 1290, 585, 11060, 1372, 1907, 13501, 11473, 16422, 517, 267, 460, 2572, 7108, 179, 10515, 3414, 1], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

A new cancer vaccine may teach the immune system to "see" abnormal cells
{'input_ids': [251, 543, 17698, 24831, 742, 4487, 107, 22203, 3659, 109, 267, 123, 3220, 179, 111, 562, 471, 10123, 532, 17617, 1], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

{'▁long': 492, '▨◠▦▱◬': 14563, '▁◆▴◧◓◕◪': 29518, '◬▱◈◬': 7264, '▁memories': 10578, '▁deserved': 21500, '▁◈▩◒◭▦▩◳◧◓': 13188, '▁□◓◧◍◪◞◣◓': 30514, '▁▫▴◒▷◗◞': 30915, '▁◞◠◚◠◒◠': 12427, '▁◎▴◞◠▸': 5494, '▁◦▱▫▪▵': 15163, '▁◀◫▱◕◫◞◠◳◠◓': 23821, '▁◫▱▨': 1193, '▁bull': 10128, '◎▪◒▫▪◎▵': 7965, 'ah': 3125, '▁veil': 27946, '▁transmission': 22229, '▁▨◓◗◞▫◠▱': 26390, '▁◠◉◬': 12498, '▁◉◠▱◬◒▪▽◂◓': 8378, '

Сохраним полученный новый токенизатор локально: 

In [ ]:
new_tokenizer.save_pretrained("./Model/new_tokenizer")

('./Model/new_tokenizer/tokenizer_config.json',
 './Model/new_tokenizer/special_tokens_map.json',
 './Model/new_tokenizer/tokenizer.json')

### Подготовка данных для обучения модели

Загрузим обученный ранее токенизатор:


In [ ]:
tokenizer = AutoTokenizer.from_pretrained("./Model/new_tokenizer")

Проверим токенизатор на паре предложений из тренировочной части датасета:

In [ ]:
example_num = 7
tokenizer(dataset['train'][example_num]['src'], dataset['train'][example_num]['dst'])

Для разных моделей может понадобиться разный подход в настройке перевода. Например для моделей семества `google-t5` необходим специальный преффикс. Реализуем это:

In [ ]:
print(model_checkpoint)

if model_checkpoint in ["google-t5/t5-small", "google-t5/t5-base", "google-t5/t5-larg", "google-t5/t5-3b", "google-t5/t5-11b"]:
    print("This is a T5 model")
    prefix = "translate Zetan to English: "
else:
    print("This is not a T5 model")
    prefix = ""

Прежде чем передать в модель тренировочнуй часть датасета необходимо его предварительно обработать. Напишем функцию которая подготовит датасет в которой:

- примеры из датасета будут подаваться на вход `tokenizer` с аргументом `truncation=True`. Это гарантирует, что входные данные, длина которых превышает ту, которую может обработать выбранная модель, будут усечены до максимальной длины, принимаемой моделью.

- Дополнение коротких последовательностей (Padding) будет реализованно позже (в коллаторе данных).

In [ ]:
# Установка максимальной длины входной и выходной последовательности
max_input_length = max_length
max_target_length = max_length

source_lang = "zetan"
target_lang = "en"

def preprocess_function(examples):
    inputs = [prefix + ex[source_lang] for ex in examples["translation"]]
    targets = [ex[target_lang] for ex in examples["translation"]]
    model_inputs = tokenizer(inputs, max_length=max_input_length, truncation=True)

    # Setup the tokenizer for targets
    with tokenizer.as_target_tokenizer():
        labels = tokenizer(targets, max_length=max_target_length, truncation=True)

    model_inputs["labels"] = labels["input_ids"]
    return model_inputs

Проверим функцию на нескольких примерах из датасета:

In [ ]:
preprocess_function(dataset['train'][:2])

# Обучение модели

Загрузим необученную модель (модель со случайно сгенерированными весами):

In [35]:
from transformers import AutoConfig, T5ForConditionalGeneration

config = AutoConfig.from_pretrained(model_checkpoint)
model = T5ForConditionalGeneration(config)

config.json:   0%|          | 0.00/1.21k [00:00<?, ?B/s]

# Дополнительные материалы

По основной задаче:

- [Training your own tokenizer from scratch](https://github.com/huggingface/notebooks/blob/main/examples/tokenizer_training.ipynb)
- [Translation fine-tuning example](https://github.com/huggingface/notebooks/blob/main/examples/translation.ipynb)

По проблемам с контейнером:

- [Jupyter notebook executions turn grey in Visual studio code](https://stackoverflow.com/questions/73481249/jupyter-notebook-executions-turn-grey-in-visual-studio-code)

# Тестовая часть

Используется для отладки кода и проведения экспериментов.